src/it/resources/expected/cpp_dependent_interface/cpp-headers/dependent_interface.hpp

```diff
@@ -0,0 +1,21 @@
+// AUTOGENERATED FILE - DO NOT MODIFY!
+// This file was generated by Djinni from cpp_dependent_interface.djinni
+
+#pragma once
+
+#include "interface_dependency.hpp"
+#include <memory>
+#include <optional>
+
+class DependentInterface {
+public:
+    virtual ~DependentInterface() {}
+
+    virtual void method_taking_interface_dependency(const dropbox::oxygen::nn_shared_ptr<::InterfaceDependency> & dep) = 0;
+
+    virtual void method_taking_optional_interface_dependency(const std::shared_ptr<::InterfaceDependency> & dep) = 0;
+
+    virtual dropbox::oxygen::nn_shared_ptr<::InterfaceDependency> method_returning_interface_dependency() = 0;
+
+    virtual std::shared_ptr<::InterfaceDependency> method_returning_optional_interface_dependency() = 0;
```

Comment:
Can you help me understand this part? My naive understanding is that that `method_taking_optional_interface_dependency` and `method_returning_optional_interface_dependency` should take/return a `std::optional` and not just a plain `shared_ptr`?

Reply:
`std::shared_ptr` is an optional type already in the sense that it can point to null. That's actually also the reason that `nn_shared_ptr` is useful in djinni: To explicitly state at compile time that a pointer is not null or in other words not optional. It's handy if we don't want to check our interface types for null wherever we use them.

An `optional<my_interface>` param is always translated to `std::shared_ptr<MyInterface>`. A non-optional `my_interface` param is translated to `std::shared_ptr<MyInterface>` as well, except if we use the `--cpp-nn-header` option. In this case, with the fix in this PR, it's now correctly translated to `nn_shared_ptr<MyInterface>`.